text/0000-introduce empty ranges.md

```diff
@@ -0,0 +1,443 @@
+
+Another problematic example of the current implementation is the response to any of the parameters being `Nan`.
+
+```pony
+  let nan = F64 = F64(0) / F64(0)
+  let r =  Range[F64](0, 10, nan)
+  for i in Range(0, 10) do
+    try env.out.print(r.next()?.string()) end
+  end
+```
+
+```
+0
+-nan(ind)
+-nan(ind)
+-nan(ind)
+-nan(ind)
+-nan(ind)
+-nan(ind)
+-nan(ind)
+-nan(ind)
+-nan(ind)
+```
+
+`NaN` is not a value and can therefore not possibly lie within any valid range. The same is true when one of the bound parameters is `NaN`. Under this RFC Range iterators that use `NaN` in any parameter are *empty*. Pony currently also considers those Ranges *infinite* were the `min, max` bounds are finite and identical but the `step` parameter is `+-Inf`. These cases too would be treated as empty since no iteration is necessary to advance from `min` to `max` regardless of the magnitude of `step`.
+
+# Detailed design
+
+## Current Implementation
+
+Currently, a Range is considered infinite if either 1) the `step` is `0`, any of `min`, `max`, or `step` are `NaN`, `+Inf` or `-Inf`, or 2) if no progress can be made from `min` to `max` due to the sign of the `step`. Here is the corresponding relevant portion of the current range.pony implementation:
+
+> ```pony
+> class Range[A: (Real[A] val & Number) = USize] is Iterator[A]
+>   ..
+> let _min: A
+>   let _max: A
+>   let _inc: A
+>   let _forward: Bool
+>   let _infinite: Bool
+>   var _idx: A
+> 
+>   new create(min: A, max: A, inc: A = 1) =>
+>     _min = min
+>     _max = max
+>     _inc = inc
+>     _idx = min
+>     _forward = (_min < _max) and (_inc > 0)
+>     let is_float_infinite =
+>       iftype A <: FloatingPoint[A] then
+>         _min.nan() or _min.infinite()
+>           or _max.nan() or _max.infinite()
+>           or _inc.nan() or _inc.infinite()
+>       else
+>         false
+>       end
+>     _infinite =
+>       is_float_infinite
+>         or ((_inc == 0) and (min != max))    // no progress
+>         or ((_min < _max) and (_inc < 0)) // progress into other directions
+>         or ((_min > _max) and (_inc > 0)) 
+> ```
+
+## Proposed reclassifications
+
+This RFC partitions the set of currently *infinite* `Range` cases into those now considered *empty*, and a very small set of those that remain *infinite*. To discuss this discrimination, we first list three criteria. All three must be met for a currently *infinite* Range to *not* now be reclassified as empty -- if either of these criteria is violated, the Range is *empty*; if all three are met, a currently *infinite* Range remains so under this proposal.
+
+  **Criterion 1** - progress from `min` to `max` must be possible. This necessitates the "no-progress expression" (discussion below) to be `false`.
+  
+  **Criterion 2** - the iterator that realizes the progress must produce *finite* values that lie within `[min, max)`.
+  
+  **Criterion 3** - none of the parameters `min, max, step` can be float `NaN`. (discussion below)
```

Comment:
```pony
use "collections"

actor Main
  new create(env: Env) =>
    let p_inf: F64 = F64.max_value() + F64.max_value()
    env.out.print(Range[F64](0, p_inf, 1).is_infinite().string())
```

Yes there are such currently infinite Ranges. Currently Pony makes a Range *infinite* if any of the bounds is infinite, so the above will print `true`. This Range remains *infinite* under this RFC because it passes Criterion 2. It does so since our Range/ interval (math.) is upper bound-open, only finite values lie within `[x, +inf)`.

Reply:
I was just about to disagree and state that there are infinite values in `[0, +inf)` however upon more consideration you are correct since the infinite exists on the open bound.